openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -501,6 +516,46 @@ def _delete_cookie(request, response):
         )
 
 
+def log_request_user_changes(request):
+    """
+    Instrument the request object so that we log changes to the `user` attribute. This is done by
+    changing the `__class__` attribute of the request object to point to a new class we created
+    on the fly which is exactly the same as the underlying request class but with an override for
+    the `__setattr__` function to catch the attribute chages.
+    """
+    original_user = getattr(request, 'user', None)
+
+    class SafeSessionRequestWrapper(request.__class__):
+        """
+        A wrapper class for the request object.
+        """
+
+        def __setattr__(self, name, value):
+            nonlocal original_user
+            if name == 'user' and hasattr(value, 'id'):
```

Comment:
Are we confident that we don't care about user being set to something without an id?

Reply:
That's a good call, it's probably worth logging.  For this particular problem, it's not relevant because the `id` is what the safe_sessions middleware looks at.